docs/cluster_template.md

```diff
@@ -41,8 +43,57 @@ awsRegion: eu-west-1
 
 Running `kops toolbox template` replaces the placeholders in the template by values and generates the file output.yaml, which can then be used to replace the desired cluster configuration with `kops replace -f cluster.yaml`.
 
+Note: when creating a cluster desired configuration template, you can
 
-Note:
-When creating a cluster desired configuration template, you can 
 - use `kops get k8s-cluster.example.com -o yaml > cluster-desired-config.yaml` to create the cluster desired configuration file (see [cli/kops_get.md](cli/kops_get.md)). The values in this file are defined in [cli/cluster_spec.md](cli/cluster_spec.md).
 - replace values by placeholders in that file to create the template.
+
+### Templates
+
+The `--template` command line option can point to either a specific file or a directory with a collection of templates. An example usage would be;
+
+```shell
+$ kops toolbox template --values dev.yaml --template cluster.yaml --template instance_group_directory
+```
+
+The cluster.yaml *(your main cluster spec for example)* would be written first followed by any templates found in the instance_group_directory directory. Note the toolbox will automatically add YAML separators between the documents for you.
+
+### Snippets
+
+The toolbox template also supports the reuse or break up of code blocks into snippets directories. By passing a `--snippets PATH` to a directory holding templates;
+
+```shell
+$ kops toolbox template --values dev.yaml --template cluster.yaml --template instancegroups --snippets snippets
+```
+
+The example below assumes you have placed the appropriate files i.e. *(nodes.json, master.json etc)* in to the snippets directory. Note, the namespace of the snippets are flat and always the basename() of the file path; so `snippets/components/docker.options` is still referred to as 'docker.options'.
+
+```YAML
+apiVersion: kops/v1alpha2
+kind: Cluster
+metadata:
+  name: {{ .environment }}.{{ .dns_zone }}
+spec:
+  docker:
+    {{ include "docker" . | indent 4 }}
+  additionalPolicies:
+    master: |
+      {{ include "masters.json" . | indent 6 }}
+    node: |
+      {{ include "nodes.json" . | indent 6 }}
+```
+
+### Template Functions
+
+The entire set of https://github.com/Masterminds/sprig functions are available within the templates for you. Note if you want to use the 'defaults' functions switch off the verification check on the command line by `--fail-on-missing=false`;
+
+```YAML
+image: {{ default $image $node.image }}
+machineType: {{ default $instance $node.machine_type }}
+maxSize: {{ default "10" $node.max_size }}
+minSize: {{ default "1" $node.min_size }}
```

Comment:
Sorry to resurrect an old PR, but where does `$node` come from in this case?

Reply:
Arrh .. yes .. I probably shouldn't have included this .. It was a copy and paste from our deployment repo, where we using a common template for all instancegroup .. so our environment file would have something like 

```YAML
node_pools:
- name: compute
  min_size: 2
- name: ingress
  instance: t2.medium
```

```YAML
{{- $base := . -}}
{{- $dns_zone := .dns_zone -}}
{{- $enable_bastion := .enable_bastion -}}
{{- $env := env "ENVIRONMENT" -}}
{{- $image := .coreos_version -}}
{{- $instance := .compute_size -}}
{{- $size := .compute_root -}}
{{- $subnets := .compute_subnets -}}
{{- range $index, $node := .node_pools -}}
{{- if $index }}
---
{{ end }}apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  labels:
    kops.k8s.io/cluster: {{ $env }}.{{ $dns_zone }}
  name: {{ $node.name }}
spec:
  {{- if (and (eq $enable_bastion false) ($node.security_groups)) }}
  additionalSecurityGroups:
  {{- range $node.security_groups }}
  - {{ . }}
  {{- end }}
  {{- end }}
  cloudLabels:
    {{- if or (not $node.cloud_labels) (not (hasKey $node.cloud_labels "role")) }}
    role: {{ $node.name }}
    {{- end }}
    {{- if $node.cloud_labels }}
    {{- range $name, $value := $node.cloud_labels }}
    {{ $name }}: {{ $value }}
    {{- end }}
    {{- end }}
  image: {{ default $image $node.image }}
  machineType: {{ default $instance $node.machine_type }}
  maxSize: {{ default "10" $node.max_size }}
  minSize: {{ default "1" $node.min_size }}
  nodeLabels:
    {{- if or (not $node.node_labels) (not (hasKey $node.node_labels "role")) }}
    role: {{ $node.name }}
    {{- end }}
    {{- if $node.node_labels }}
    {{- range $name, $value := $node.node_labels }}
    {{ $name }}: {{ $value }}
    {{- end }}
    {{- end }}
  role: {{ default "Node" $node.role }}
  rootVolumeSize: {{ default $size $node.root_size }}
  subnets:
  {{- $networks := default $subnets $node.subnets }}
  {{- range $networks }}
  {{- if typeIs "map[string]interface {}" . }}
  - {{ .name }}
  {{- else }}
  - {{ . }}
  {{- end }}
  {{- end }}
  {{- if $node.taints }}
  taints:
  {{- range $node.taints }}
  - {{ .}}
  {{- end }}
  {{- end }}
{{- end }}
```

Reply:
I thought it might be something like that, and that example was what I was after. Thanks for sharing!